### PR TITLE
Slice ep

### PR DESCRIPTION
**Pull request by jshipton**
Changes to make Euler-Poincare form of nonlinear advection for u work for vertical slice. Tested (minimally) and with example file.

### CODE REVIEW

**Comment by jshipton** [comment]
Just realised this needs a lot of comments!

File: dcore/advection.py [hunk 1]
```diff
@@ -202,3 +202,94 @@ def apply(self, x_in, x_out):
...
+
+    def apply(self, x_in, x_out):
+        self.u0.assign(x_in)
+        self.usolver.solve()
+        x_out.assign(self.u1)
+
+
+class SUPGAdvection(Advection):
+
+    def __init__(self, state, V, direction=None, supg_params=None):
```
Comment (jshipton) [inline dcore/advection.py]: 'direction' is supposed to specify the direction in which we're not applying SUPG - so the entry for tau is set to zero for that direction. I've just realised I also need to change the surface measure, depending on 'direction'.

File: dcore/advection.py [hunk 2]
```diff
@@ -202,3 +202,117 @@ def apply(self, x_in, x_out):
...
+        self.u0.assign(x_in)
+        self.usolver.solve()
+        x_out.assign(self.u1)
+
+
+class SUPGAdvection(Advection):
+    """
+    An SUPG advection scheme that applies DG upwinding in the direction
+    specified by the direction arg.
+
```
Comment (colinjcotter) [inline dcore/advection.py]: I think that this needs a better name, since it is not pure SUPG (has upwind DG too).

**Comment by jshipton** [comment]
Can this go in now please? (If it passes tests obvs.) The changes are:
1) Euler Poincare form of advection for vertical slice model (previously only worked for shallow water)
2) Hybrid SUPG-DG upwind advection for continuous or partially continuous fields
3) Diffusion (interior penalty method)
4) sponge layer for compressible slice
5) examples files for dense bubble, rising bubble and non hydrostatic mountain test cases
6) tests for nonlinear vertical slice code, hybrid advection scheme and diffusion scheme
Sorry this merge has got so unwieldy - all the elements were required to test that the nonlinear velocity advection was working for the slice code.
The remaining issues are:
1) parallel issues (can't get comm from extruded mesh and mountain example not parallel safe) I know how to fix both of these but would like to do it in a separate branch.
2) I have removed the surface term in the nonlinear advection code as it seems to mess up the falling bubble - requires more testing
3) tidier hydrostatic initialisation (in another branch)

**Comment by colinjcotter** [comment]
Have you seen the hydros branch? There I changed over to a geopotential
formulation, and added a corresponding hydrostatic pressure solver.

cheers
--cjc

On 29 June 2016 at 18:40, jshipton <notifications@github.com> wrote:

> Can this go in now please? (If it passes tests obvs.) The changes are:
> 1) Euler Poincare form of advection for vertical slice model (previously
> only worked for shallow water)
> 2) Hybrid SUPG-DG upwind advection for continuous or partially continuous
> fields
> 3) Diffusion (interior penalty method)
> 4) sponge layer for compressible slice
> 5) examples files for dense bubble, rising bubble and non hydrostatic
> mountain test cases
> 6) tests for nonlinear vertical slice code, hybrid advection scheme and
> diffusion scheme
> Sorry this merge has got so unwieldy - all the elements were required to
> test that the nonlinear velocity advection was working for the slice code.
> The remaining issues are:
> 1) parallel issues (can't get comm from extruded mesh and mountain example
> not parallel safe) I know how to fix both of these but would like to do it
> in a separate branch.
> 2) I have removed the surface term in the nonlinear advection code as it
> seems to mess up the falling bubble - requires more testing
> 3) tidier hydrostatic initialisation (in another branch)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229431858>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF0q8m0ymP5PcERd7K692U0A7y3BnuNEks5qQq4SgaJpZM4IgdYT>
> .
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916

**Comment by jshipton** [comment]
Yes, I saw there was a branch - is it ready to merge? Do I need to adapt if for the slice?

________________________________

**Comment by colinjcotter** [comment]
It's nearly ready, I just need to see results tomorrow.

For slice, you'll need to use the geopotential form of the equations too.

cheers
--cjc

On 29 June 2016 at 20:07, jshipton <notifications@github.com> wrote:

> Yes, I saw there was a branch - is it ready to merge? Do I need to adapt
> if for the slice?
>
> ________________________________
> From: Colin J Cotter <notifications@github.com>
> Sent: 29 June 2016 19:43:44
> To: firedrakeproject/gusto
> Cc: Shipton, Jemma; Author
> Subject: Re: [firedrakeproject/gusto] Slice ep (#38)
>
>
> Have you seen the hydros branch? There I changed over to a geopotential
> formulation, and added a corresponding hydrostatic pressure solver.
>
> cheers
> --cjc
>
> On 29 June 2016 at 18:40, jshipton <notifications@github.com> wrote:
>
> > Can this go in now please? (If it passes tests obvs.) The changes are:
> > 1) Euler Poincare form of advection for vertical slice model (previously
> > only worked for shallow water)
> > 2) Hybrid SUPG-DG upwind advection for continuous or partially continuous
> > fields
> > 3) Diffusion (interior penalty method)
> > 4) sponge layer for compressible slice
> > 5) examples files for dense bubble, rising bubble and non hydrostatic
> > mountain test cases
> > 6) tests for nonlinear vertical slice code, hybrid advection scheme and
> > diffusion scheme
> > Sorry this merge has got so unwieldy - all the elements were required to
> > test that the nonlinear velocity advection was working for the slice
> code.
> > The remaining issues are:
> > 1) parallel issues (can't get comm from extruded mesh and mountain
> example
> > not parallel safe) I know how to fix both of these but would like to do
> it
> > in a separate branch.
> > 2) I have removed the surface term in the nonlinear advection code as it
> > seems to mess up the falling bubble - requires more testing
> > 3) tidier hydrostatic initialisation (in another branch)
> >
> > -
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229431858>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe/AF0q8m0ymP5PcERd7K692U0A7y3BnuNEks5qQq4SgaJpZM4IgdYT
> >
> > .
> >
>
>
>
> --
> http://www.imperial.ac.uk/people/colin.cotter
>
> www.cambridge.org/9781107663916
>
> -
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub<
> https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229450221>,
> or mute the thread<
> https://github.com/notifications/unsubscribe/AGjpyi_4sN0OiR4WJMv-jH2zoQR7t-0Zks5qQrzggaJpZM4IgdYT
> >.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229456829>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF0q8n0keGrqO7NTVMRY57J187N9B5sgks5qQsJkgaJpZM4IgdYT>
> .
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916

**Comment by jshipton** [comment]
OK!


Can I merge slice_EP if it passes tests?

________________________________

**Comment by colinjcotter** [comment]
Yes, if you correct the spelling of Penalty :)

--cjc

On 29 June 2016 at 20:19, jshipton <notifications@github.com> wrote:

> OK!
>
>
> Can I merge slice_EP if it passes tests?
>
> ________________________________
> From: Colin J Cotter <notifications@github.com>
> Sent: 29 June 2016 20:17:36
>
> To: firedrakeproject/gusto
> Cc: Shipton, Jemma; Author
> Subject: Re: [firedrakeproject/gusto] Slice ep (#38)
>
> It's nearly ready, I just need to see results tomorrow.
>
> For slice, you'll need to use the geopotential form of the equations too.
>
> cheers
> --cjc
>
> On 29 June 2016 at 20:07, jshipton <notifications@github.com> wrote:
>
> > Yes, I saw there was a branch - is it ready to merge? Do I need to adapt
> > if for the slice?
> >
> > ________________________________
> > From: Colin J Cotter <notifications@github.com>
> > Sent: 29 June 2016 19:43:44
> > To: firedrakeproject/gusto
> > Cc: Shipton, Jemma; Author
> > Subject: Re: [firedrakeproject/gusto] Slice ep (#38)
> >
> >
> > Have you seen the hydros branch? There I changed over to a geopotential
> > formulation, and added a corresponding hydrostatic pressure solver.
> >
> > cheers
> > --cjc
> >
> > On 29 June 2016 at 18:40, jshipton <notifications@github.com> wrote:
> >
> > > Can this go in now please? (If it passes tests obvs.) The changes are:
> > > 1) Euler Poincare form of advection for vertical slice model
> (previously
> > > only worked for shallow water)
> > > 2) Hybrid SUPG-DG upwind advection for continuous or partially
> continuous
> > > fields
> > > 3) Diffusion (interior penalty method)
> > > 4) sponge layer for compressible slice
> > > 5) examples files for dense bubble, rising bubble and non hydrostatic
> > > mountain test cases
> > > 6) tests for nonlinear vertical slice code, hybrid advection scheme and
> > > diffusion scheme
> > > Sorry this merge has got so unwieldy - all the elements were required
> to
> > > test that the nonlinear velocity advection was working for the slice
> > code.
> > > The remaining issues are:
> > > 1) parallel issues (can't get comm from extruded mesh and mountain
> > example
> > > not parallel safe) I know how to fix both of these but would like to do
> > it
> > > in a separate branch.
> > > 2) I have removed the surface term in the nonlinear advection code as
> it
> > > seems to mess up the falling bubble - requires more testing
> > > 3) tidier hydrostatic initialisation (in another branch)
> > >
> > > -
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> > https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229431858
> >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe/AF0q8m0ymP5PcERd7K692U0A7y3BnuNEks5qQq4SgaJpZM4IgdYT
> > >
> > > .
> > >
> >
> >
> >
> > --
> > http://www.imperial.ac.uk/people/colin.cotter
> >
> > www.cambridge.org/9781107663916
> >
> > -
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub<
> > https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229450221
> >,
> > or mute the thread<
> >
> https://github.com/notifications/unsubscribe/AGjpyi_4sN0OiR4WJMv-jH2zoQR7t-0Zks5qQrzggaJpZM4IgdYT
> > >.
> >
> > -
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229456829>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe/AF0q8n0keGrqO7NTVMRY57J187N9B5sgks5qQsJkgaJpZM4IgdYT
> >
> > .
> >
>
>
>
> --
> http://www.imperial.ac.uk/people/colin.cotter
>
> www.cambridge.org/9781107663916
>
> -
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub<
> https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229459639>,
> or mute the thread<
> https://github.com/notifications/unsubscribe/AGjpymNGuEJB0mxZWsWQK3h5cM8thywsks5qQsTQgaJpZM4IgdYT
> >.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/gusto/pull/38#issuecomment-229460033>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF0q8r0vcUgqhXwLNTfG6FiwsSU5VtSqks5qQsUqgaJpZM4IgdYT>
> .
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916

**Pull request merged**